dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/weld/WebSphereCDIDeploymentImpl.java

```diff
@@ -565,38 +566,93 @@ public WebSphereBeanDeploymentArchive getBeanDeploymentArchive(Class<?> beanClas
 
     @Override
     @Trivial
-    //This method is called a lot
     public WebSphereBeanDeploymentArchive getBeanDeploymentArchiveFromClass(Class<?> clazz) {
-        // Note this method looks for a BDA which has the given class in it, **whether the class is a bean or not**
-        // We needed this so that we can find the correct bean manager for non-beans like servlets
+
         WebSphereBeanDeploymentArchive wbda = classBDAMap.get(clazz);
 
         if (wbda == null) {
-            for (WebSphereBeanDeploymentArchive bda : orderedBDAs) {
-                if (bda.getAllClazzes().contains(clazz.getName())) {
 
-                    Class<?> bdaClazz = null;
-                    try {
-                        bdaClazz = CDIUtils.loadClass(bda.getClassLoader(), clazz.getName());
-                    } catch (CDIException e) {
-                        // TODO Auto-generated catch block
-                        // Do you need FFDC here? Remember FFDC instrumentation and @FFDCIgnore
-                        // https://websphere.pok.ibm.com/~liberty/secure/docs/dev/API/com.ibm.ws.ras/com/ibm/ws/ffdc/annotation/FFDCIgnore.html
-                        e.printStackTrace();
-                    }
-                    if (bdaClazz == clazz) {
+            for (WebSphereBeanDeploymentArchive searchBDA : orderedBDAs) {
+                if (doesBDAContainClazz(searchBDA, clazz)) {
+                    wbda = searchBDA;
+                    break;
+                }
+            }
 
-                        wbda = bda;
-                        classBDAMap.put(clazz, bda);
-                        break;
+            if (wbda != null) {
+                if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                    Tr.debug(tc, "mapping class: " + clazz.getName() + " " + System.identityHashCode(clazz) + " to bda: " + wbda.toString());
+                }
+
+                classBDAMap.putIfAbsent(clazz, wbda);
+
+                //Performance optimization, since most of the time a class will be in one BDA, it makes sense to check
+                //if its an EJB module here and record that now.
+                if (wbda.getType() == ArchiveType.EJB_MODULE) {
+                    if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                        Tr.debug(tc, "mapping class: " + clazz.getName() + " " + System.identityHashCode(clazz) + " to EJB_MODULE bda: " + wbda.toString());
                     }
+                    classBDAMapFavouringEjb.putIfAbsent(clazz, wbda);
+                }
+            }
+        }
+
+        return wbda;
+    }
+
+    @Trivial
+    @Override
+    public WebSphereBeanDeploymentArchive getBeanDeploymentArchiveFromClassFavouringEJB(Class<?> clazz) {
+
+        WebSphereBeanDeploymentArchive wbda = classBDAMap.get(clazz);
+
```

Comment:
Checking the wrong map here I think

Reply:
I'm surprised we didn't get a test failure here. We may have just got unlucky and called `getBeanDeploymentArchiveFromClassFavouringEJB` before `getBeanDeploymentArchiveFromClass` for the test EJB class.

If you haven't done so, it's probably worth reverting the implementation changes and ensuring that the test you've added fails.

Reply:
I'm not sure what prompts `getBeanDeploymentArchiveFromClass` to be called. I guess if we never call it before `getBeanDeploymentArchiveFromClassFavouringEJB` then we wouldn't see a problem with this mistake.